### ISSUE TITLE
Connection to MQTT over Websockets with HTTP Proxy hangs indefinitely

### ISSUE THREAD

**Issue by JeYdevoir** [aws/aws-iot-device-sdk-java-v2#57]
Confirm by changing [ ] to [x] below:
- [x] I've searched for [previous similar issues](https://github.com/aws/aws-iot-device-sdk-java-v2/issues) and didn't find any solution

**Platform/OS/Hardware/Device**
OS: Mac High Sierra
Device: Android 8.0
4in WVGA Nexus S Emulator

**Describe the question**
I'm trying to connect to MQTT over Websockets with a proxy through my machine as a test for our android app, but once I call "connection.connect()" it just seems to not do anything. When I run the PubSub.java sample and configure with websockets, it executes correctly and I see the activity in my proxy, so I'm thinking there's something I'm missing here or a weird limitation in Android I haven't discovered.

If I remove the websockets configuration, it connects fine and I'm able to publish to a topic without any issues. The device itself has its wifi settings configured for the proxy as well, and I see traffic for google play just fine.

Edit: I tried removing the proxy and just using websockets, and I get the same behavior - it hangs indefinitely.

When I execute this code:
```
     private void connectWithNewStuff() throws Exception {
        LOG.debug("/*-+-*/ New connection start!");
        MqttMessage lwt = new MqttMessage(getString(R.string.messaging_service_lwt_topic),
                getString(R.string.messaging_service_aws_lwt).getBytes());

        try (EventLoopGroup eventLoopGroup = new EventLoopGroup(1);
             HostResolver resolver = new HostResolver(eventLoopGroup);
             ClientBootstrap bootstrap = new ClientBootstrap(eventLoopGroup, resolver);
             AwsIotMqttConnectionBuilder builder = AwsIotMqttConnectionBuilder
                     .newMtlsBuilder(getCertificate(), getPrivateKey())
                     .withBootstrap(bootstrap)
                     .withWill(lwt, QualityOfService.AT_MOST_ONCE, false)
                     .withClientId("02:00:00:44:55:46")
                     .withCertificateAuthority(getRootCa())
                     .withCleanSession(true)
                     .withConnectionEventCallbacks(callbacks)
                     .withEndpoint("[ats endpoint].iot.us-east-1.amazonaws.com")) {
            LOG.debug("/*-+-*/ Builder Established");
            HttpProxyOptions proxy = new HttpProxyOptions();
            proxy.setHost("192.168.1.217");
            proxy.setPort(8888);
            builder.withWebsockets(true)
                    .withWebsocketSigningRegion("us-east-1")
                    .withWebsocketProxyOptions(proxy);
            try (MqttClientConnection connection = builder.build()) {
                LOG.debug("/*-+-*/ Connection built~");
                CompletableFuture<Boolean> connected = connection.connect();
                LOG.debug("/*-+-*/ Connection set to completable future.");
                try {
                    boolean sessionPresent = connected.get();
                    LOG.debug("/*-+-*/ Connected to " + (!sessionPresent ? "new" : "existing")
                            + " session!");
                } catch (Exception ex) {
                    throw new RuntimeException(
                            "/*-+-*/ Exception occurred during new connect stuff", ex);
                }
                LOG.debug("/*-+-*/ Connection established!");
            }
        }
    }
```

I only see the following logs before it stops entirely. I've let it sit there for a few minutes, but nothing happens.
```
D/com.tracelink.core.service.MessagingService: clientId = 02:00:00:44:55:46
D/FA: Connected to remote service
D/com.amazonaws.request: Initializing the regions with default regions
V/FA: Processing queued up service tasks: 6
D/com.tracelink.core.service.MessagingService: /*-+-*/ New connection start!
W/View: dispatchProvideAutofillStructure(): not laid out, ignoring
I/AssistStructure: Flattened final assist data: 4560 bytes, containing 1 windows, 17 views
D/com.tracelink.core.service.MessagingService: /*-+-*/ Builder Established
D/com.tracelink.core.service.MessagingService: /*-+-*/ Connection built~
V/FA: Inactivity, disconnecting from the service
I/zygote: Do full code cache collection, code=116KB, data=99KB
I/zygote: After code cache collection, code=102KB, data=58KB
I/zygote: Do partial code cache collection, code=122KB, data=79KB
I/zygote: After code cache collection, code=122KB, data=79KB
    Increasing code cache capacity to 512KB
D/com.tracelink.core.service.MessagingService: /*-+-*/ Connection set to completable future.
```

**Comment by justinboswell** [comment]
I will see if I can repro this myself via emulator. Should be straightforward.

**Comment by justinboswell** [comment]
In the meantime, can you capture and attach a log at TRACE level? at startup, call ```software.amazon.awssdk.crt.Log.initLoggingToStderr(LogLevel.Trace)```. This will log our internals to logcat.

**Comment by justinboswell** [comment]
I am able to reproduce this. For me, it appears to be a lack of AWS Credentials to sign the websocket request. I'm looking deeper into it with our signing engineer.

**Comment by JeYdevoir** [comment]
Here are the logs you asked for.

```
D/com.tracelink.core.service.MessagingService: /*-+-*/ Attempted connection
D/task-scheduler: [9015d970]  - id=0x95545180: Running attempt_connection task with <Running> status
D/socket: [9015d970]  - id=0xa47aef00 fd=63: initializing with domain 0 and type 0
    [9015d970]  - id=0xa47aef00 fd=63: setting socket options to: keep-alive 0, keep idle 0, keep-alive interval 0, keep-alive probe count 0.
W/socket: [9015d970]  - id=0xa47aef00 fd=63: setsockopt() for NO_SIGNAL failed with errno 92. If you are having SIGPIPE signals thrown, you may want to install a signal trap in your application layer.
D/socket: [9015d970]  - id=0xa47aef00 fd=63: beginning connect.
    [9015d970]  - id=0xa47aef00 fd=63: connecting to endpoint 169.254.169.254:80.
V/socket: [9015d970]  - id=0xa47aef00 fd=63: connection pending waiting on event-loop notification or timeout.
V/event-loop: [9015d970]  - id=0x909f3940: subscribing to events on fd 63
V/socket: [9015d970]  - id=0xa47aef00 fd=63: scheduling timeout task for 967376768740.
V/event-loop: [9015d970]  - id=0x909f3940: scheduling task 0x92768f00 in-thread for timestamp 967376768740
D/task-scheduler: [9015d970]  - id=0x92768f00: Scheduling (null) task for future execution at time 967376768740
V/event-loop: [9015d970]  - id=0x909f3940: detected more scheduled tasks with the next occurring at 1999851000, using timeout of 1999.
    [9015d970]  - id=0x909f3940: waiting for a maximum of 1999 ms
V/dns: [8ff7f970]  - static, resolving 169.254.169.254, unsolicited resolve count 1
D/dns: [8ff7f970]  - static: resolving host 169.254.169.254
    [8ff7f970]  - static: resolved record: 169.254.169.254
    [8ff7f970]  - static: resolved record: 169.254.169.254
V/dns: [8ff7f970]  - static: updating expiry for 169.254.169.254 for host 169.254.169.254 to 1588857661435756000
    [8ff7f970]  - static: updating expiry for 169.254.169.254 for host 169.254.169.254 to 1588857661435756000
    [8ff7f970]  - static: remaining record count for host 0
    [8ff7f970]  - static: remaining record count for host 1
V/dns: [8ff7f970]  - static, resolving 169.254.169.254, unsolicited resolve count 2
D/dns: [8ff7f970]  - static: resolving host 169.254.169.254
D/dns: [8ff7f970]  - static: resolved record: 169.254.169.254
    [8ff7f970]  - static: resolved record: 169.254.169.254
V/dns: [8ff7f970]  - static: updating expiry for 169.254.169.254 for host 169.254.169.254 to 1588857662437026000
    [8ff7f970]  - static: updating expiry for 169.254.169.254 for host 169.254.169.254 to 1588857662437026000
    [8ff7f970]  - static: remaining record count for host 0
    [8ff7f970]  - static: remaining record count for host 1
V/event-loop: [9015d970]  - id=0x909f3940: wake up with 0 events to process.
    [9015d970]  - id=0x909f3940: running scheduled tasks.
D/task-scheduler: [9015d970]  - id=0x92768f00: Running (null) task with <Running> status
V/socket: [9015d970]  - task_id=0x92768f00: timeout task triggered, evaluating timeouts.
E/socket: [9015d970]  - id=0xa47aef00 fd=63: timed out, shutting down.
V/event-loop: [9015d970]  - id=0x909f3940: un-subscribing from events on fd 63
    [9015d970]  - id=0x909f3940: scheduling task 0x909fe1d0 in-thread for timestamp 0
D/task-scheduler: [9015d970]  - id=0x909fe1d0: Scheduling epoll_event_loop_unsubscribe_cleanup task for immediate execution
D/socket: [9015d970]  - id=0xa47aef00 fd=63: closing
E/socket: [9015d970]  - id=0xa47aef00 fd=-1: connection failure
D/channel-bootstrap: [9015d970]  - id=0x927689c0: client connection on socket 0xa47aef00 completed with error 1048.
    [9015d970]  - id=0x927689c0: recording bad address 169.254.169.254.
I/dns: [9015d970]  - id=0x909ed2f0: recording failure for record 169.254.169.254 for 169.254.169.254, moving to bad list
D/dns: [9015d970]  - static: purging address 169.254.169.254 for host 169.254.169.254 from the cache due to cache eviction or shutdown
V/channel-bootstrap: [9015d970]  - id=0x927689c0: releasing socket 0xa47aef00 either because we already have a successful connection or because it errored out.
D/socket: [9015d970]  - id=0xa47aef00 fd=-1: closing
E/channel-bootstrap: [9015d970]  - id=0x927689c0: Connection failed with error_code 1048.
E/http-connection: [9015d970]  - static: Client connection failed with error 1048 (AWS_IO_SOCKET_TIMEOUT).
W/connection-manager: [9015d970]  - id=0x92699a80: Failed to obtain new connection from http layer, error 1048(socket operation timed out.)
D/connection-manager: [9015d970]  - id=0x92699a80: Failing excess connection acquisition with error code 1048
    [9015d970]  - id=0x92699a80: snapshot - state=1, held_connection_count=0, pending_acquire_count=0, pending_connect_count=0, vended_connection_count=0, open_connection_count=0, ref_count=1
W/AuthCredentialsProvider: [9015d970]  - id=0x92768c00: Instance metadata provider failed to acquire a connection, error code 1048(socket operation timed out.)
I/AuthCredentialsProvider: [9015d970]  - (id=0x926f3a30) Credentials provider chain ending query on chain member 2 with invalid credentials
D/AuthCredentialsProvider: [9015d970]  - (id=0x909c9ac0) Cached credentials provider next refresh time set to 1867379406740
    [9015d970]  - (id=0x909c9ac0) Cached credentials provider was unable to source credentials on refresh
    [9015d970]  - (id=0x909c9ac0) Cached credentials provider notifying pending queries of new credentials
E/AuthSigning: [9015d970]  - (id=0x909ed520) Credentials Provider provided no credentials
D/aws-c-common: [9015d970]  - aws_jni_get_thread_env returned detached, attaching
W/AuthCredentialsProvider: [9015d970]  - (id=0x92768c00) IMDS credentials provider failed to query instance role credentials
W/connection-manager: [9015d970]  - id=0x0: Failed to completed connection acquisition with error_code 1048(socket operation timed out.)
```

**Comment by justinboswell** [comment]
Yeah, ok, so you have the same issue as my repro, which boils down to not having any AWS credentials available, which are required to sign websocket requests. The best production option would be to use Cognito to vend STS tokens, which will provide per-device ephemeral credentials, which is enough to sign the HTTP request.

Alternatively, you can use other credentials mechanisms (e.g. environment variables or credentials file, though on android we don't have explicit support for faking a home directory yet. I'm looking into how we might support that). Let me dig a bit further and see what we can do in the short term.

**Comment by justinboswell** [comment]
I was able to get this working locally through a squid proxy:
* Set ```AWS_ACCESS_KEY_ID```/```AWS_SECRET_ACCESS_KEY``` environment variables to a valid set of credentials. You could equally pass in Cognito or other credentials.
* Ensure that the root CA is provided to the connection builder, and no client MTLS is configured
* Provide the proxy host and port to the connection builder

**Comment by justinboswell** [comment]
Oh, also, is your proxy running on your local machine? If so, you will need to put in the address as 10.0.2.2, which is what the emulator maps to 127.0.0.1 on the host, and ensure that your proxy is bound to 127.0.0.1.

**Comment by JeYdevoir** [comment]
I've been using the machine's IP as a proxy IP/connection point, and it works fine that way. No need for localhost shenanigans. 

I'll be trying your solution now, I'll let you know if I run into any other issues. Thank you!

**Comment by JeYdevoir** [comment]
I've added those two fields on my .bash_profile with valid credentials, transformed the constructor to not use MTLS, but I'm still getting the same thing. Is there anything in particular I have to do with android studio or the emulator once I update my profile?

My updated code for the constructor looks something like this:
```
try (EventLoopGroup eventLoopGroup = new EventLoopGroup(1);
             HostResolver resolver = new HostResolver(eventLoopGroup);
             ClientBootstrap bootstrap = new ClientBootstrap(eventLoopGroup, resolver);
             AwsIotMqttConnectionBuilder builder = AwsIotMqttConnectionBuilder
                     .newDefaultBuilder()
                     .withBootstrap(bootstrap)
                     .withWill(lwt, QualityOfService.AT_MOST_ONCE, false)
                     .withClientId("02:00:00:44:55:46")
                     .withCertificateAuthority(getRootCa())
                     .withCleanSession(true)
                     .withConnectionEventCallbacks(callbacks)
                     .withEndpoint("a2tth08d02ufib-ats.iot.us-east-1.amazonaws.com")) {
            LOG.debug("/*-+-*/ Builder Established");
            HttpProxyOptions proxy = new HttpProxyOptions();
            proxy.setHost("192.168.1.217");
            proxy.setPort(8888);
            builder.withWebsockets(true)
                    .withWebsocketSigningRegion("us-east-1")
                    .withWebsocketProxyOptions(proxy);
            try (MqttClientConnection connection = builder.build()) {
                LOG.debug("/*-+-*/ Connection built~");
                CompletableFuture<Boolean> connected = connection.connect();
                LOG.debug("/*-+-*/ Connection set to completable future.");
                try {
                    boolean sessionPresent = connected.get();
```

**Comment by justinboswell** [comment]
You need credentials on the device, not on the host machine, see my edits above for how to set the process' environment variables on device. Environment variables are not propagated to the emulator.

Other than that, that connection code looks good.

**Comment by JeYdevoir** [comment]
That seemed to do the trick! I don't have an Android 8.0 device handy to test this physically, but would it need anything similar to the emulator? I don't believe we'll be giving our customers AWS keys for their devices so they need to be able to connect with just a keystore (certificate and private key). How would that work, exactly? Or does this problem not come up in testing before on a physical device?

**Comment by justinboswell** [comment]
Cognito is the service for this, it allows anonymous identity creation and binding to IAM roles. Work with your TAMs if you need more service info, I'm not a service expert, but the basics are:
* Create an Identity Pool and assign an Unauthenticated IAM role. 
  * In your case, that policy should just be enough to reach IoT. 
* Do something akin to the samples here: https://docs.amplify.aws/lib/auth/getting-started/q/platform/android, either by using the android AWS SDK, or just using the HTTP layer in the IoT Device SDK to make the REST call to get a credential pair.

**Comment by JeYdevoir** [comment]
Two things;
I think the explanation needs to be elaborated on just a little bit. I'm afraid I'm not following really. I went through that link and used the automated setup process but something doesn't feel right here. I'm not sure what it means to create an identity pool or how to provide unauthenticaed IAM role.

If the automated setup on that link is the right path, for my second thing:
Is this required to be part of the repo?
```
"CognitoUserPool": {
            "Default": {
                "PoolId": "XX-XXXX-X_abcd1234",
                "AppClientId": "XXXXXXXX",
                "AppClientSecret": "XXXXXXXXX",
                "Region": "XX-XXXX-X"
            }
        }
```
I'm not okay with committing credential keys into the repo.

**Comment by justinboswell** [comment]
You will need to contact your account managers for support. Unfortunately, I can only support the SDK, service guidance is best handled by the service teams.

**Comment by JeYdevoir** [comment]
The more I look at it, the more this seems like a much bigger change for the application than anticipated. There's no way to connect through websockets/proxy without additional AWS Setup than the certificate/private key? Based on what I found here, there needs to be configuration changes made on our AWS account itself that the application can query/pull from somehow? https://docs.aws.amazon.com/cognito/latest/developerguide/identity-pools.html 

I know you're only supporting the SDK, but what _can_ you tell me about connecting through websockets? It sounds like we might not have to use Cognito if we're not feasibly able to go through the setup for our purposes.

**Comment by justinboswell** [comment]
You can also do X.509 authentication to pull credentials. We have an implementation of X.509 auth, but no Java bindings just yet. Please work with your TAM if you would prefer to authenticate that way. We can try to get those bindings added sooner rather than later.

In any event, the IoT authentication options are available here:
https://docs.aws.amazon.com/iot/latest/developerguide/client-authentication.html

Once you pick which method you want to use and can pull credentials to the device, you'll be able to sign websocket connections and punch through proxies.

**Comment by JeYdevoir** [comment]
Okay, we got the information for the X.509 stuff and the example in PubSub.java seems straightforward, but I'm still facing issues and the logs I'm seeing are not very descriptive. Do you think you could tell me if I am setting this up incorrectly, or some steps I could try?

Or is this what you mean when you say "it's not ready yet"? (Note: I'm not introducing proxy yet, just websockets for now.)

There seems to be one log in particular that claims "ca_file is null" but I know for a fact I'm passing it in, so I'm not sure what it means exactly.

Here's how I have the X.509 provider set up (the rest of the code to connect should look the same as the examples provided above):
```
try (TlsContextOptions x509TlsOptions = TlsContextOptions
                    .createWithMtlsFromPath("/sdcard/Download/46cert.txt",
                            "/sdcard/Download/46key.txt")
                    .withCertificateAuthorityFromPath(null, "/sdcard/Download/AwsRootCA1.txt")) {
                LOG.debug("/*-+-*/ Created x509 Tls Context Options");
                try (ClientTlsContext x509TlsContext = new ClientTlsContext(x509TlsOptions)) {
                    X509CredentialsProvider.X509CredentialsProviderBuilder x509builder = new X509CredentialsProvider.X509CredentialsProviderBuilder()
                            .withClientBootstrap(bootstrap)
                            .withTlsContext(x509TlsContext)
                            .withEndpoint("[ats endpoint].iot.us-east-1.amazonaws.com")
                            .withRoleAlias("jlamendola")
                            .withThingName("02:00:00:44:55:46");
                    try (X509CredentialsProvider provider = x509builder.build()) {
                        builder.withWebsocketCredentialsProvider(provider);
                    }
                }
            }
            LOG.debug("/*-+-*/ Added x509 Credentials Provider");
```

And here are the logs I'm experiencing:
```
D/com.tracelink.core.service.MessagingService: /*-+-*/ New connection start!
I/dns: [91f7a970]  - id=0x9018ce30: Initializing default host resolver with 8 max host entries.
I/channel-bootstrap: [91f7a970]  - id=0xa6f90100: Initializing client bootstrap with event-loop group 0x901958e0
D/com.tracelink.core.service.MessagingService: /*-+-*/ Builder Established
D/com.tracelink.core.service.MessagingService: /*-+-*/ Added websockets and proxy configuration!
D/com.tracelink.core.service.MessagingService: /*-+-*/ Created x509 Tls Context Options
D/tls-handler: [91f7a970]  - ctx: Certificate and key have been set, setting them up now.
I/connection-manager: [91f7a970]  - id=0xa6fdbc00: Successfully created
D/tls-handler: [91f7a970]  - ctx: Certificate and key have been set, setting them up now.
D/com.tracelink.core.service.MessagingService: /*-+-*/ Added x509 Credentials Provider
E/tls-handler: [91f7a970]  - ctx: configuration error NULL pointer encountered (Error encountered in ../../../../../aws-common-runtime/s2n/tls/s2n_config.c line 437)
    [91f7a970]  - Failed to set ca_file (null)
I/connection-manager: [91f7a970]  - id=0xa6fdbc00: release
    [91f7a970]  - id=0xa6fdbc00: ref count now zero, starting shut down process
    [91f7a970]  - id=0xa6fdbc00: manager release, failing 0 pending acquisitions
D/connection-manager: [91f7a970]  - id=0xa6fdbc00: snapshot - state=2, held_connection_count=0, pending_acquire_count=0, pending_connect_count=0, vended_connection_count=0, open_connection_count=0, ref_count=0
I/connection-manager: [91f7a970]  - id=0xa6fdbc00: Destroying self
D/AuthCredentialsProvider: [91f7a970]  - Credentials providers shutdown complete
V/AuthCredentialsProvider: [91f7a970]  - CrtCredentialsProvider.onShutdownComplete
D/channel-bootstrap: [91f7a970]  - id=0xa6f90100: releasing bootstrap reference
    [91f7a970]  - id=0xa6f90100: destroying
V/channel-bootstrap: [91f7a970]  - ClientBootstrap.onShutdownComplete
I/event-loop: [8fcd0970]  - id=0x90195900: Destroying event_loop
    [8fcd0970]  - id=0x90195900: Stopping event-loop thread.
V/event-loop: [8fcd0970]  - id=0x90195900: Scheduling task 0xa6f9fa4c cross-thread for timestamp 0
    [8fcd0970]  - id=0x90195900: Waking up event-loop thread
V/event-loop: [8fdcd970]  - id=0x90195900: wake up with 1 events to process.
    [8fdcd970]  - id=0x90195900: activity on fd 61, invoking handler.
    [8fdcd970]  - id=0x90195900: notified of cross-thread tasks to schedule
    [8fdcd970]  - id=0x90195900: processing cross-thread tasks
    [8fdcd970]  - id=0x90195900: task 0xa6f9fa4c pulled to event-loop, scheduling now.
D/task-scheduler: [8fdcd970]  - id=0xa6f9fa4c: Scheduling epoll_event_loop_stop task for immediate execution
V/event-loop: [8fdcd970]  - id=0x90195900: running scheduled tasks.
D/task-scheduler: [8fdcd970]  - id=0xa6f9fa4c: Running epoll_event_loop_stop task with <Running> status
V/event-loop: [8fdcd970]  - id=0x90195900: no more scheduled tasks using default timeout.
D/event-loop: [8fdcd970]  - id=0x90195900: exiting main loop
V/event-loop: [8fdcd970]  - id=0x90195900: un-subscribing from events on fd 61
    [8fdcd970]  - id=0x90195900: scheduling task 0x90dbf9d0 in-thread for timestamp 0
D/task-scheduler: [8fdcd970]  - id=0x90dbf9d0: Scheduling epoll_event_loop_unsubscribe_cleanup task for immediate execution
D/com.tracelink.core.service.MessagingService: /*-+-*/ New connection END END END!
E/com.tracelink.core.service.MessagingService: /*-+-*/ Error occurred during new connection stuff: software.amazon.awssdk.crt.CrtRuntimeException: TlsContext.tls_ctx_new: Failed to create new aws_tls_ctx (aws_last_error: AWS_IO_TLS_CTX_ERROR(1033), Failed to create tls context)
    software.amazon.awssdk.crt.CrtRuntimeException: TlsContext.tls_ctx_new: Failed to create new aws_tls_ctx (aws_last_error: AWS_IO_TLS_CTX_ERROR(1033), Failed to create tls context)
        at software.amazon.awssdk.crt.io.TlsContext.tlsContextNew(Native Method)
        at software.amazon.awssdk.crt.io.TlsContext.<init>(TlsContext.java:34)
        at software.amazon.awssdk.crt.io.ClientTlsContext.<init>(ClientTlsContext.java:36)
        at software.amazon.awssdk.iot.AwsIotMqttConnectionBuilder.build(AwsIotMqttConnectionBuilder.java:411)
        at com.tracelink.core.service.MessagingService.connectWithNewStuff(MessagingService.java:580)
        at com.tracelink.core.service.MessagingService.connectToMqtt(MessagingService.java:528)
        at com.tracelink.core.service.MessagingService.establishConnection(MessagingService.java:503)
        at com.tracelink.core.service.MessagingService.connect(MessagingService.java:416)
        at com.tracelink.core.service.-$$Lambda$QGvInWxgQaux-XIMYYPhle8KTFw.run(Unknown Source:2)
        at com.tracelink.core.service.MessagingService.lambda$wrap$0(MessagingService.java:436)
        at com.tracelink.core.service.-$$Lambda$MessagingService$knrcCIKmsULdFEyFPCA7CjogF58.call(Unknown Source:2)
        at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
        at io.reactivex.Single.subscribe(Single.java:3096)
        at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
        at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:571)
        at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
        at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
```

**Comment by justinboswell** [comment]
What happens if you read the cert/key/CA files yourself, and use the non-path (```createWithMtls```/```withCertificateAuthority```) APIs? Are you able to read them? From the looks of it, the code looks right, I'm wondering if there's file read permissions problems.

**Comment by JeYdevoir** [comment]
Yeah you were right, there was definitely a reading issue (that I'm mad I didn't see sooner), I've fixed that and now I'm seeing something else that I'm not sure how to decipher. Any hints?

```
E/http-connection: [90140970]  - static: Client connection failed with error 1029 (AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE).
W/connection-manager: [90140970]  - id=0x92624800: Failed to obtain new connection from http layer, error 1029(TLS (SSL) negotiation failed)
D/connection-manager: [90140970]  - id=0x92624800: Failing excess connection acquisition with error code 1029
    [90140970]  - id=0x92624800: snapshot - state=1, held_connection_count=0, pending_acquire_count=0, pending_connect_count=0, vended_connection_count=0, open_connection_count=0, ref_count=1
W/connection-manager: [90140970]  - id=0x0: Failed to complete connection acquisition with error_code 1029(TLS (SSL) negotiation failed)
W/AuthCredentialsProvider: [90140970]  - id=0x90541380: X509 provider failed to acquire a connection, error code 1029(TLS (SSL) negotiation failed)
E/AuthCredentialsProvider: [90140970]  - Failed to parse IoT Core response as Json document.
W/AuthCredentialsProvider: [90140970]  - (id=0x90541380) X509 credentials provider failed to query credentials
E/AuthSigning: [90140970]  - (id=0x9051df50) Credentials Provider provided no credentials
```

**Comment by bretambrose** [comment]
Hello,

The per-account credentials provider endpoint is different than your per-account iot thing connection endpoint and it looks like you may be using the thing endpoint here.

To get the right endpoint for credentials, use this cli command:

aws iot describe-endpoint --endpoint-type iot:CredentialProvider

**Comment by JeYdevoir** [comment]
Didn't know that either - thank you so much for that information. 

However, the plot thickens; Getting a 403 response now. Endpoint was a variable I wasn't sure exactly how to use and was just guessing, so I guess now I should address "RoleAlias" - what should this value be? Would an incorrect RoleAlias have anything to do with the 403?

Here's the trace:
```
V/channel: [8fed8970]  - id=0x95635c00: sending read message of size 265, from slot 0x95672b00 to slot 0x9567eea0 with handler 0xa6f9e384.
V/http-connection: [8fed8970]  - id=0xa6f9e380: Begin processing message of size 265.
V/http-stream: [8fed8970]  - id=0x95630cc0: Incoming response status: 403 (Forbidden).
    [8fed8970]  - id=0x95630cc0: Incoming header: content-type: application/json
D/AuthCredentialsProvider: [8fed8970]  - (id=0x914ec100) X509 credentials provider query received http status code 403
V/http-stream: [8fed8970]  - id=0x95630cc0: Incoming header: content-length: 53
    [8fed8970]  - id=0x95630cc0: Incoming header: date: Tue, 19 May 2020 14:39:00 GMT
    [8fed8970]  - id=0x95630cc0: Incoming header: x-amzn-RequestId: 94b20d1d-5912-9a7a-0706-43384e28e628
    [8fed8970]  - id=0x95630cc0: Incoming header: x-amzn-ErrorType: AccessDeniedException
    [8fed8970]  - id=0x95630cc0: Main header block done.
    [8fed8970]  - id=0x95630cc0: Incoming body: 53 bytes received.
V/AuthCredentialsProvider: [8fed8970]  - (id=0x914ec100) X509 credentials provider received 53 response bytes
D/http-stream: [8fed8970]  - id=0x95630cc0: Client request complete, response status: 403 (Forbidden).
V/http-stream: [8fed8970]  - id=0x95630cc0: Stream refcount released, 1 remaining.
E/AuthCredentialsProvider: [8fed8970]  - Failed to parse IoT Core response as Json document.
W/AuthCredentialsProvider: [8fed8970]  - (id=0x914ec100) X509 credentials provider failed to query credentials
E/AuthSigning: [8fed8970]  - (id=0x9259be80) Credentials Provider provided no credentials
```

**Comment by bretambrose** [comment]
In my imperfect understanding, the role alias is the bridge between your IoT permissions (is this thing allowed to fetch credentials?) and traditional AWS IAM roles and permissions (required for the IoT credentials service to invoke STS for you).  

There's quite a bit of work in setting it up.  When I added Java bindings to the X509 provider and updated the PubSub sample, I followed this tutorial: https://aws.amazon.com/blogs/security/how-to-eliminate-the-need-for-hardcoded-aws-credentials-in-devices-by-using-the-aws-iot-credentials-provider/ 

Be careful though because it has some properties which will mess you up if you follow it blindly.  In particular, at the very minimum you will want to rework the credentials-fetching policy associated with the thing's certificate to not require the thing to be connected (section 5 of the tutorial) since that's the endgoal we're trying to accomplish here.

**Comment by JeYdevoir** [comment]
While we'd like to avoid IAM altogether, I'm giving it a shot to make sure if it comes down to it this will work if we need to use it, but I'm still getting the 403 I've described above even though i'm pretty sure I've set everything up correctly.

1: I've created an IAM Role with two policies;
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "VisualEditor0",
            "Effect": "Allow",
            "Action": [
                "sts:DecodeAuthorizationMessage",
                "sts:GetCallerIdentity"
            ],
            "Resource": "*"
        },
        {
            "Sid": "VisualEditor1",
            "Effect": "Allow",
            "Action": "sts:*",
            "Resource": "arn:aws:iam::*:role/*"
        }
    ]
}
```
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iot:*"
            ],
            "Resource": "*"
        }
    ]
}
```
2: I've created a Role Alias in IOT referring to this IAM role.
3: I've created this policy in IOT:
```
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": "iot:AssumeRoleWithCertificate",
      "Resource": "*"
    }
  ]
}
```
4: I've attached this policy to the Thing's certificate. 

According to the resource you've provided here, that should be the requirements to get this to function, correct? Is there something I'm missing? Is there something I can get into further detail that might be causing the issue?

**Comment by JeYdevoir** [comment]
So, one thing I realized, you never explicitly said to add "-ats" to the credentials endpoint, and I took it off. Now I'm getting a difference exception but seems like progress, but I'm definitely stuck again.
```
V/channel: [9027f970]  - id=0xb41b4dc0: acquired message 0x8ff09d00 of length 0 from pool 0x92739130. Requested size was 4294967295
V/tls-handler: [9027f970]  - id=0x902a6600: Bytes read 296
V/channel: [9027f970]  - id=0xb41b4dc0: sending read message of size 296, from slot 0x90298fa0 to slot 0x8ff50340 with handler 0xa4773184.
V/http-connection: [9027f970]  - id=0xa4773180: Begin processing message of size 296.
V/http-stream: [9027f970]  - id=0x9261ef80: Incoming response status: 400 (Bad Request).
    [9027f970]  - id=0x9261ef80: Incoming header: content-type: application/json
D/AuthCredentialsProvider: [9027f970]  - (id=0x90292700) X509 credentials provider query received http status code 400
V/http-stream: [9027f970]  - id=0x9261ef80: Incoming header: content-length: 77
    [9027f970]  - id=0x9261ef80: Incoming header: date: Thu, 21 May 2020 20:52:44 GMT
    [9027f970]  - id=0x9261ef80: Incoming header: x-amzn-RequestId: eb1d1f8b-1e72-a40d-6821-80098953d8b2
    [9027f970]  - id=0x9261ef80: Incoming header: x-amzn-ErrorType: MisconfiguredRoleException
    [9027f970]  - id=0x9261ef80: Main header block done.
    [9027f970]  - id=0x9261ef80: Incoming body: 77 bytes received.
V/AuthCredentialsProvider: [9027f970]  - (id=0x90292700) X509 credentials provider received 77 response bytes
D/http-stream: [9027f970]  - id=0x9261ef80: Client request complete, response status: 400 (Bad Request).
V/http-stream: [9027f970]  - id=0x9261ef80: Stream refcount released, 1 remaining.
E/AuthCredentialsProvider: [9027f970]  - Failed to parse IoT Core response as Json document.
W/AuthCredentialsProvider: [9027f970]  - (id=0x90292700) X509 credentials provider failed to query credentials
E/AuthSigning: [9027f970]  - (id=0x9558e400) Credentials Provider provided no credentials
```

**Comment by bretambrose** [comment]
Well, I see "Incoming header: x-amzn-ErrorType: MisconfiguredRoleException" in the response, but it's not easy to tell exactly what that means for you.

Can you try fully relaxing all of your policies to complete wildcards?  If that doesn't lead to success, then there's some kind of missing reference or association between the AWS resources.  If it does lead to success, then we just need to figure out the right policy scoping.

**Comment by JeYdevoir** [comment]
Changing the policies to simple wildcards did not fix the issue. What details should I go over more thoroughly for you to deduce what references/associations I'm missing?

**Comment by bretambrose** [comment]
Honestly, given how opaque request errors are, I think the best bet would be to engage your TAM/SA contact(s) and work directly with someone within the IoT service (using the request ids that you see in the logs).  We (SDKs) have no insight into internal failures beyond what shows up in the response and reasoning against what's in the tutorials.

**Comment by JeYdevoir** [comment]
I actually managed to see where it was not configured correctly and fixed it, but now I'm getting the same exception as I was originally (shown below). So I believe the credentials provider is correctly set up now, but when we try to connect, it still doesn't like the credentials for some reason. When I look at the logs, I see the "Amz-Security-Token" value populated with a seemingly legitimate string, so we know it's doing something right in that regard. 

What would be the next step in debugging?

Here is the logs surrounding the error (the error is at the very bottom):
```
V/socket-handler: [90241970]  - id=0xa6fa0e00: read 2609 from socket
V/channel: [90241970]  - id=0x95671c00: sending read message of size 2609, from slot 0x956432e0 to slot 0x91583ec0 with handler 0x91584900.
W/tls-handler: [90241970]  - id=0x91584900: negotiation failed with error Bad message encountered (Error encountered in ../../../../../aws-common-runtime/s2n/tls/s2n_handshake_io.c line 1001)
D/channel-bootstrap: [90241970]  - id=0x90a88960: tls negotiation result 1029 on channel 0x95671c00
V/channel: [90241970]  - id=0x95671c00: channel shutdown task is scheduled
    [90241970]  - id=0x95671c00: scheduling task with wrapper task id 0x95671d04.
V/event-loop: [90241970]  - id=0x926e64a0: scheduling task 0x95671d04 in-thread for timestamp 0
D/task-scheduler: [90241970]  - id=0x95671d04: Scheduling channel_shutdown task for immediate execution
D/channel: [90241970]  - id=0x95671c00: Channel shutdown is already pending, not scheduling another.
V/channel: [90241970]  - id=0x95671c00: acquired message 0x91593c40 of length 0 from pool 0x955b2d40. Requested size was 11079
V/socket: [90241970]  - id=0xa6fa5f40 fd=66: read of -1
    [90241970]  - id=0xa6fa5f40 fd=66: read would block
V/socket-handler: [90241970]  - id=0x956432f0: total read on this tick 5305
    [90241970]  - id=0xa6fa0e00: out of data to read on socket. Waiting on event-loop notification.
V/socket: [90241970]  - id=0xa6fa5f40 fd=66: is writable
    [90241970]  - id=0xa6fa5f40 fd=66: processing write requests.
    [90241970]  - id=0xa6fa5f40 fd=66: processing write requests, invoked by the event-loop
V/event-loop: [90241970]  - id=0x926e64a0: running scheduled tasks.
D/task-scheduler: [90241970]  - id=0x95671d04: Running channel_shutdown task with <Running> status
D/channel: [90241970]  - id=0x95671c00: beginning shutdown process
V/channel: [90241970]  - id=0x95671c00: shutting down slot 0x956432e0 (the first one) in the read direction
    [90241970]  - id=0x95671c00: shutting down slot 0x956432e0, with handler 0xa6fa0e00 in read direction with error code 1029
V/socket-handler: [90241970]  - id=0xa6fa0e00: shutting down read direction with error_code 1029
D/channel: [90241970]  - id=0x95671c00: handler 0xa6fa0e00 shutdown in read dir completed.
D/tls-handler: [90241970]  - id=0x91584900: Shutting down read direction with error code 1029
D/channel: [90241970]  - id=0x95671c00: handler 0x91584900 shutdown in read dir completed.
V/event-loop: [90241970]  - id=0x926e64a0: scheduling task 0x95671c14 in-thread for timestamp 0
D/task-scheduler: [90241970]  - id=0x95671c14: Scheduling (null) task for immediate execution
V/event-loop: [90241970]  - id=0x926e64a0: detected more scheduled tasks with the next occurring at 0, using timeout of 0.
    [90241970]  - id=0x926e64a0: waiting for a maximum of 0 ms
    [90241970]  - id=0x926e64a0: wake up with 1 events to process.
    [90241970]  - id=0x926e64a0: activity on fd 66, invoking handler.
V/socket: [90241970]  - id=0xa6fa5f40 fd=66: is writable
    [90241970]  - id=0xa6fa5f40 fd=66: processing write requests.
    [90241970]  - id=0xa6fa5f40 fd=66: processing write requests, invoked by the event-loop
V/event-loop: [90241970]  - id=0x926e64a0: running scheduled tasks.
D/task-scheduler: [90241970]  - id=0x95671c14: Running (null) task with <Running> status
D/tls-handler: [90241970]  - id=0x91584900: Shutting down write direction
V/channel: [90241970]  - id=0x95671c00: acquired message 0x95614ee0 of length 0 from pool 0x955b2d40. Requested size was 7
    [90241970]  - id=0x95671c00: sending write message of size 7, from slot 0x91583ec0 to slot 0x956432e0 with handler 0xa6fa0e00.
V/socket-handler: [90241970]  - id=0xa6fa0e00: writing message of size 7
V/socket: [90241970]  - id=0xa6fa5f40 fd=66: processing write requests.
    [90241970]  - id=0xa6fa5f40 fd=66: processing write requests, called from aws_socket_write
    [90241970]  - id=0xa6fa5f40 fd=66: dequeued write request of size 7, remaining to write 7
    [90241970]  - id=0xa6fa5f40 fd=66: send written size 7
    [90241970]  - id=0xa6fa5f40 fd=66: remaining write request to write 0
    [90241970]  - id=0xa6fa5f40 fd=66: write request completed
V/socket-handler: [90241970]  - static: write of size 7, completed on channel 0x95671c00
D/channel: [90241970]  - id=0x95671c00: handler 0x91584900 shutdown in write dir completed.
V/socket-handler: [90241970]  - id=0xa6fa0e00: shutting down write direction with error_code 1029
D/socket: [90241970]  - id=0xa6fa5f40 fd=66: closing
V/event-loop: [90241970]  - id=0x926e64a0: un-subscribing from events on fd 66
    [90241970]  - id=0x926e64a0: scheduling task 0x9562d810 in-thread for timestamp 0
D/task-scheduler: [90241970]  - id=0x9562d810: Scheduling epoll_event_loop_unsubscribe_cleanup task for immediate execution
V/channel: [90241970]  - id=0x95671c00: scheduling task with wrapper task id 0xa6fa0e54.
V/event-loop: [90241970]  - id=0x926e64a0: scheduling task 0xa6fa0e54 in-thread for timestamp 0
D/task-scheduler: [90241970]  - id=0xa6fa0e54: Scheduling socket_handler_close task for immediate execution
V/event-loop: [90241970]  - id=0x926e64a0: detected more scheduled tasks with the next occurring at 0, using timeout of 0.
    [90241970]  - id=0x926e64a0: waiting for a maximum of 0 ms
    [90241970]  - id=0x926e64a0: wake up with 0 events to process.
    [90241970]  - id=0x926e64a0: running scheduled tasks.
D/task-scheduler: [90241970]  - id=0x9562d810: Running epoll_event_loop_unsubscribe_cleanup task with <Running> status
    [90241970]  - id=0xa6fa0e54: Running socket_handler_close task with <Running> status
D/channel: [90241970]  - id=0x95671c00: handler 0xa6fa0e00 shutdown in write dir completed.
V/event-loop: [90241970]  - id=0x926e64a0: scheduling task 0x95671c14 in-thread for timestamp 0
D/task-scheduler: [90241970]  - id=0x95671c14: Scheduling (null) task for immediate execution
V/event-loop: [90241970]  - id=0x926e64a0: detected more scheduled tasks with the next occurring at 0, using timeout of 0.
    [90241970]  - id=0x926e64a0: waiting for a maximum of 0 ms
    [90241970]  - id=0x926e64a0: wake up with 0 events to process.
    [90241970]  - id=0x926e64a0: running scheduled tasks.
D/task-scheduler: [90241970]  - id=0x95671c14: Running (null) task with <Running> status
D/channel: [90241970]  - id=0x95671c00: during shutdown, canceling task 0x91584918
V/event-loop: [90241970]  - id=0x926e64a0: cancelling task 0x91584918
D/task-scheduler: [90241970]  - id=0x91584918: Running tls_timeout task with <Canceled> status
D/channel-bootstrap: [90241970]  - id=0x90a88960: channel 0x95671c00 shutdown with error 1029.
E/http-connection: [90241970]  - static: Client connection failed with error 1029 (AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE).
E/websocket-setup: [90241970]  - id=0xa6f87a00: Websocket setup failed to establish HTTP connection, error 1029 (AWS_IO_TLS_ERROR_NEGOTIATION_FAILURE).
V/mqtt-client: [90241970]  - id=0xb41b6800: Channel has been shutdown with error code 1029
```

**Comment by bretambrose** [comment]
Can you post the full configuration of your AwsIotMqttConnectionBuilder from declaration up through build() call?

**Comment by JeYdevoir** [comment]
Sure.

```
/**
     * Method.
     */
    private void connectWithNewStuff() throws Exception {
        software.amazon.awssdk.crt.Log.initLoggingToStderr(Log.LogLevel.Trace);
        LOG.debug("/*-+-*/ New connection start!");
        MqttMessage lwt = new MqttMessage(getString(R.string.messaging_service_lwt_topic),
                getString(R.string.messaging_service_aws_lwt).getBytes());
        try (EventLoopGroup eventLoopGroup = new EventLoopGroup(1);
             HostResolver resolver = new HostResolver(eventLoopGroup);
             ClientBootstrap bootstrap = new ClientBootstrap(eventLoopGroup, resolver);
             AwsIotMqttConnectionBuilder builder = AwsIotMqttConnectionBuilder
                     .newMtlsBuilderFromPath("/sdcard/Download/46cert.txt",
                             "/sdcard/Download/46key.txt")
                     .withBootstrap(bootstrap)
                     .withWill(lwt, QualityOfService.AT_MOST_ONCE, false)
                     .withClientId("02:00:00:44:55:46")
                     .withCertificateAuthorityFromPath(null, "/sdcard/Download/AwsRootCA1.txt")
                     .withCleanSession(true)
                     .withConnectionEventCallbacks(callbacks)
                     .withEndpoint("[iot ats endpoint].iot.us-east-1.amazonaws.com")) {
            LOG.debug("/*-+-*/ Builder Established");
            HttpProxyOptions proxy = new HttpProxyOptions();
            proxy.setHost("192.168.1.217");
            proxy.setPort(8888);
            builder.withWebsockets(true)
                    .withWebsocketSigningRegion("us-east-1");
            //        .withWebsocketProxyOptions(proxy);
            LOG.debug("/*-+-*/ Added websockets and proxy configuration!");
            try (TlsContextOptions x509TlsOptions = TlsContextOptions
                    .createWithMtlsFromPath("/sdcard/Download/46cert.txt",
                            "/sdcard/Download/46key.txt")
                    .withCertificateAuthorityFromPath(null, "/sdcard/Download/AwsRootCA1.txt")) {
                LOG.debug("/*-+-*/ Created x509 Tls Context Options");
                try (ClientTlsContext x509TlsContext = new ClientTlsContext(x509TlsOptions)) {
                    X509CredentialsProvider.X509CredentialsProviderBuilder x509builder = new X509CredentialsProvider.X509CredentialsProviderBuilder()
                            .withClientBootstrap(bootstrap)
                            .withTlsContext(x509TlsContext)
                            .withEndpoint("[credentials endpoint].credentials.iot.us-east-1.amazonaws.com")
                            .withRoleAlias("jlamendola-sit-device")
                            .withThingName("02:00:00:44:55:46");
                    try (X509CredentialsProvider provider = x509builder.build()) {
                        builder.withWebsocketCredentialsProvider(provider);
                    }
                }
            }
            LOG.debug("/*-+-*/ Added x509 Credentials Provider");
            try (MqttClientConnection connection = builder.build()) {
                LOG.debug("/*-+-*/ Connection built~");
                CompletableFuture<Boolean> connected = connection.connect();
           ...
```

**Comment by bretambrose** [comment]
Just to clarify: you are successfully retrieving credentials from the credentials endpoint but as soon as the sdk tries to establish the http connection for websockets, tls negotiation fails?

I just realized that the updated PubSub that showed a proxy working with the x509 provider wasn't actually published/released.  I do notice that you aren't setting proxy options in the x509 provider builder and aren't setting them on the main connection builder, so is this just running on a local device that doesn't have the final proxy setup?

Looking closer, I realized what it is: your main TLS setup is using a cert and key but you're connecting with websockets.  Your main TLS setup should be vanilla/default.  Authenticating via AWS credentials (over websockets) is a complete replacement for connecting via cert/key.  You still need cert/key for the x509 connection, but you shouldn't use them for the overall connection.

The pubsub example with x509 is a little bit deceptive because it unconditionally calls newMtlsBuilderFromPath on the mqtt connection builder, but when I run it successfully I was not supplying cert and key which makes that a null call.  Now that I've looked at this again, I do remember, experiencing what you are experiencing while debugging this, and so I stopped setting --cert and --key and --rootca when running the sample (and that's why --cert, --key, and --rootca is distinct from --x509cert, --x509key, and --x509rootca).

Here's my PubSub command line (should be able to trace exactly what is necessary from it):

--clientId
ShadowTest
--x509rootca
./AmazonRootCA1.pem
--x509cert
./ShadowTest.cert.pem
--x509key
./ShadowTest.private.key
--endpoint
<something>-ats.iot.us-west-2.amazonaws.com
--count
10
-w
--region
us-west-2
--x509
--x509rolealias
x509-role-alias-test
--x509endpoint
<something>.credentials.iot.us-west-2.amazonaws.com
--x509thing
ShadowTest
--proxyhost
127.0.0.1
--proxyport
3128

I will get the updated PubSub merged and released shortly (for comparison purposes).

**Comment by JeYdevoir** [comment]
We had a similar thought and I ran the PubSub.java example with all of my information, but the kicker is that i did supply the certificate and private key into the default builder but it actually worked! So I didn't think there was an issue, but I can definitely try it without the cert/key in the main setup.

And yes, I'm not enabling proxy quite yet, I'm making sure the connect over websockets as a basic concept works first and then I'll enable proxy and test that specifically. Minimizing the variables in regards to having issues and such.

Here was my setup:
```
-c
/Users/jlamendola/Documents/Certs/46cert.txt
-k
/Users/jlamendola/Documents/Certs/46key.txt
-e
[endpoint].iot.us-east-1.amazonaws.com
-t
edge-iot/hello-world
-m
"Helllooo!"
--clientId
02:00:00:44:55:46
-w
--x509
--x509cert
/Users/jlamendola/Documents/Certs/46cert.txt
--x509key
/Users/jlamendola/Documents/Certs/46key.txt
--x509rootca
/Users/jlamendola/Documents/Certs/AwsRootCA1.txt
--x509thing
02:00:00:44:55:46
--x509endpoint
[endpoint].credentials.iot.us-east-1.amazonaws.com
--x509rolealias
jlamendola-sit-device
--region
us-east-1
```

**Comment by JeYdevoir** [comment]
Oh wow, that worked! I took out the cert/private key from the initial constructor and it worked! I'll let you try to figure out how i managed to get it working from the PubSub example, but I don't think that's as important.

Thank you for the help!

**Comment by bretambrose** [comment]
That is strange but it does look like you're not supplying the rootca in the overall builder so maybe that was the important difference?

So is it working for you now?

Also, I just kicked a release of v1.1.1 of the SDK which contains the proxy support (via a dependency update to aws-crt) and updates PubSub to optionally use proxy with x509.

**Comment by JeYdevoir** [comment]
Connection through proxy worked. We need to discuss internally what the impact of IAM is on our side of things, but I think I can close this for now. Anything we would further pursue would make the most sense in a new ticket, if necessary.

Thanks again everyone for all of your help these past few weeks~
